avoid recursive exec() of QgsCredentials

jef-n opened the pull request:
followup c9e76164982 - mutex didn't only guard the cache, but also the dialog

nyalldawson commented:
@jef-n I don't understand the need for this -- and I believe it's going to bring back the original deadlock which my fix was designed to avoid.

Why would the dialog need a mutex? it's always run in the main thread

jef-n committed:
[needsbackport] apply an alternative fix for #20826

Partly reverts c9e761649, which removed the synchronizatiion of
credential requests (eg. in a project that has multiple layers from the
same postgresql database without credentials) and led to multiple
concurrent requests for the same credentials.

Some of which were silently discarded, when events processed in the
dialogs exec() event loop tried to reinvoke the dialog and caused
invalid layers.

Authentications caused by network requests can still cause this.

The credential cache is now guarded by a separate mutex.

jef-n commented:
If there are multiple layers eg. from the same postgres database requiring the same credentials, there will be multiple credential requests in parallel - but the second and following are canceled on exec() because the first request is still executing.

nyalldawson commented:
Can you please test to make sure #20826 hasn't been revived?

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -1589,6 +1589,10 @@ QgisApp::~QgisApp()
   qDeleteAll( mCustomDropHandlers );
   qDeleteAll( mCustomLayoutDropHandlers );
 
+  // replace gui based network access managers with failing only ones
```
This part should not be revived

elpaso commented:
Please have a look to my comments on https://issues.qgis.org/issues/21461

giedrioks26 commented:
this is related https://issues.qgis.org/issues/21582?

jef-n commented:
> this is related https://issues.qgis.org/issues/21582?

yes.  The reverted change cause this
